### ISSUE TITLE
How to send data to HID devices

### ISSUE THREAD

**Issue by BLesnau** [Mr-Kittyhawk/dualsense-dotnet#1]
I feel like I'm abusing the issue system on this repo, but I did not know of another way to get in touch with the creator of this repo. I was looking into doing something very similar to what this repo seems to be a starting point for. I'm interested in helping out, but am just inexperienced with doing this type of USB device / HID communication. If you're interested, I can of course chat in more detail outside of this repo, but this is where I'm at currently.

I have been researching how to interact with a game controller like a PlayStation 4 or PlayStation 5 controller. I would like to be able to determine when a button is pressed, how far a trigger is pushed down, and any other input a user could perform. The other part I would like to find out how to do is provide input to the controller. This could be something like telling the controller to rumble, or it could tell the PlayStation 5 controller to adjust the tension in its triggers, which is a new feature to that controller.

I have gotten far enough to where I think I can find the device, connect to it, and then start receiving reports with data from the controller. I also see ways to write data to the controller, but I have not tried that yet.

The main barrier I am running into is that I do not know how to parse that data that I am getting back. I am getting 64 bytes of data, but don't know what any of it is supposed to mean. Likewise, I don't know what format to use when sending data to the controller.

I know I can monitor that 64 bytes to see what changes when in press buttons on the controller, but I'm wondering if there is a good way for me to get this information without trial and error. I also don't know of a similar thing that I can do to figure out how to provide data to the controller myself.

Is there a way to get this information from the device itself, or would this information need to be provided by the controller manufacturer?

I think I've gotten pretty much as far as this repo has with just monitoring input.

**Comment by Mr-Kittyhawk** [comment]
The controller itself won't tell you how to communicate with it in a more meaningful way than it sending you HID Input reports and you sending it HID output reports back. Sony almost certanly won't publish the specifications for the contents of these reports outside of the API and documentation they give out to developers registered to receive playstation 5 devkits under NDA. The DS4Windows community has documented pretty much everything about the contents of the HID reports if you do some googling. You should be aware that the reports are slightly different when the connection to the controller is wired vs bluetooth. If you want to chat more you can bother me on the discord server linked in the readme file for the repo 😄

**Comment by BLesnau** [comment]
Thank you so much for the information, and sorry that I did not notice the Discord server. I did stumble across some of the info from the DS4Windows community, so I'm trying to decipher some of that. I suspected that there was not going to be a good way to get information from the controller itself about how to interact with it, but I just did not know for sure since I've not dug into this HID area at all up to this point. It does look like the DS4Windows community and others just needed to reverse engineer things.

Thanks again!
